### ISSUE TITLE
Simple if-else block logic with AND (`&&`) operator is not understood by TS

### ISSUE THREAD

**Issue by silvio2402** [microsoft/TypeScript#56221]
### 🔎 Search Terms

"AND operator", "if else", "condition", "types wrong"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.2.2#code/FAFwngDgpgBAsmA8gIwFYwLwwN4wgJwHsIAuGAZxHwEsA7AcxgF8BuUSWBAFQ83iTQwAPjACutACZQAZnSgS2wAMaFalGAFswAMXFK+ACgCGZbhwA0MZKbA9oASkwA+HMBjuY1aTAMBCIzAAZIEwvsiO2G4e0QD0MVaEIAAWMEb4sOJSsrTyUe5MMFAANuSwXj7+EXnRVgB0BMQsMHGF+ET4ZADkyJ2e5HiE5OTUyEVgMJ2ZMnISndVMwAtAA

### 💻 Code

```ts
type MyObj = { prop: string };
type MyType = MyObj | undefined;

const myFunc = (a: MyType, b: MyType) => {
    if (!a && !b) {
        // both are undefined
    } else if (!a) {
        b.prop; // error: 'b' is possibly 'undefined'
    }
}
```

### 🙁 Actual behavior

In the else-if block, the `b` variable is falsely assigned type `MyType`, when it should be `MyObj`, since `b` can't be undefined.

### 🙂 Expected behavior

The `b` variable should be detected as type `MyObj`.

## Workaround

The workaround is simply to restructure the if statement into two. However, I believe that TypeScript should be able to work both ways.

```ts
type MyObj = { prop: string };
type MyType = MyObj | undefined;

const myFunc = (a: MyType, b: MyType) => {
    if (!a) {
        if (!b) {
            // both are undefined
        } else {
            b.prop;
        }
    }
}
```

**Comment by RyanCavanaugh** [comment]
Narrowing works by forming a control flow graph of things that might affect a variable's type, and checking those things to see if they're syntactic forms that look like they might cause something to be narrowed. It is not a constraint-solver style thing that can handle all kinds of other combinations and permutations of logical operators, so if your logic depends on a sort of counterfactual induction, it's not going to get picked up by narrowing. Thankfully this generally corresponds with more readable code in most cases.

**Comment by fatcerberus** [comment]
[Just going to link this here like I always do when issues like this come up.](https://github.com/microsoft/TypeScript/issues/52822#issuecomment-1434989002)
